dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_memory.cc

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nvjpeg_memory.h"
+#include <nvjpeg.h>
+#include <atomic>
+#include <cassert>
+#include <iostream>
+#include <fstream>
+#include <map>
+#include <vector>
+#include <mutex>
+#include <shared_mutex>
+#include "dali/kernels/alloc.h"
+#include "dali/core/cuda_error.h"
+
+namespace dali {
+
+using kernels::memory::KernelUniquePtr;
+using kernels::AllocType;
+using kernels::memory::alloc_unique;
+using kernels::memory::Allocate;
+
+namespace nvjpeg_memory {
+
+struct Buffer {
+  Buffer(KernelUniquePtr<char> unq_ptr, AllocType type, size_t sz)
+      : ptr(std::move(unq_ptr)), alloc_type(type), size(sz) {}
+  KernelUniquePtr<char> ptr;
+  AllocType alloc_type;
```

Comment:
Do you need this filed as BufferPool has already an array that maps AllocType to buffer vector.

Reply:
Don't really need it but in case you want to use `Buffer` outside of the context of buffer pool, you'd want to know whether this is a pinned or device buffer. Added for completeness

Reply:
OK